tiny_dnn/core/kernels/avx_kernel_common.h

```diff
@@ -56,6 +56,17 @@ inline __m128d madd128_sd(__m128d a, __m128d b, __m128d c) {
 }
 #endif
 
+// Horizontally add elements of __m128 type argument
+// x = ( x3, x2, x1, x0 )
+// use _mm_cvtss_f32 if you need a float result instead of a __m128
+inline __m128 hsum128_ps(__m128 x) {
+  // twoSum = ( x3+x2, x1+x0, x3+x2, x1+x0 )
+  const __m128 twoSum = _mm_hadd_ps(x, x);
+  // fourSum = ( x3+x2+x1+x0, x3+x2+x1+x0, x3+x2+x1+x0, x3+x2+x1+x0 )
+  const __m128 fourSum = _mm_hadd_ps(twoSum, twoSum);
+  return fourSum;
+}
+
```

Comment:
Please check this page. http://stackoverflow.com/questions/6996764/fastest-way-to-do-horizontal-float-vector-sum-on-x86

It looks like this is faster.
```cpp
// in  : ( x3, x2, x1, x0 )
// out : (  -,  -,  -, x3+x2+x1+x0 )
inline __m128 hsum128_ps(__m128 x)
{
  // loDual = ( -, -, x1, x0 )
  const __m128 loDual = x;
  // hiDual = ( -, -, x3, x2 )
  const __m128 hiDual = _mm_movehl_ps(x, x);
  // sumDual = ( -, -, x1+x3, x0+x2 )
  const __m128 sumDual = _mm_add_ps(loDual, hiDual);
  // lo = ( -, -, -, x0+x2 )
  const __m128 lo = sumDual;
  // hi = ( -, -, -, x1+x3 )
  const __m128 hi = _mm_shuffle_ps(sumDual, sumDual, 0x1);
  // sum = ( -, -, -, x0+x1+x2+x3 )
  const __m128 sum = _mm_add_ss(lo, hi);
  return sum;
}
```